packages/ra-core/src/controller/details/useCreateController.ts

```diff
@@ -147,57 +148,59 @@ export const useCreateController = <
-                                          ? error
-                                          : error.message ||
+                        onSuccess:
+                            onSuccessFromSave ?
+                                onSuccessFromSave :
+                            onSuccessRef.current ?
+                                onSuccessRef.current :
+                                ({ data: newRecord }) => {
+                                    notify(
+                                        successMessage ||
+                                            'ra.notification.created',
+                                        'info',
+                                        {
+                                            smart_count: 1,
+                                        }
+                                    );
+                                    redirect(
+                                        redirectTo,
+                                        basePath,
+                                        newRecord.id,
+                                        newRecord
+                                    );
+                                },
+                        onFailure:
+                            onFailureFromSave ?
+                                onFailureFromSave :
+                            onFailureRef.current ?
+                                onFailureRef.current :
+                                error => {
+                                    notify(
+                                        typeof error === 'string' ?
+                                            error :
+                                            error.message ||
                                                 'ra.notification.http_error',
-                                      'warning',
-                                      {
-                                          _:
-                                              typeof error === 'string'
-                                                  ? error
-                                                  : error && error.message
-                                                  ? error.message
-                                                  : undefined,
-                                      }
-                                  );
-                              },
+                                        'warning',
+                                        {
+                                            _:
+                                                typeof error === 'string' ?
+                                                    error :
+                                                error && error.message ?
+                                                    error.message :
+                                                    undefined,
+                                        }
+                                    );
+                                },
```

Comment:
The code in this file gets gnarly but it shows nested ternaries in several situations